source/opt/instrument_pass.cpp

```diff
@@ -0,0 +1,951 @@
+  (*new_blk_ptr)->AddInstruction(std::move(newCallOp));
+}
+
+bool InstrumentPass::IsSameBlockOp(const Instruction* inst) const {
+  return inst->opcode() == SpvOpSampledImage || inst->opcode() == SpvOpImage;
+}
+
+void InstrumentPass::CloneSameBlockOps(
+    std::unique_ptr<Instruction>* inst,
+    std::unordered_map<uint32_t, uint32_t>* postCallSB,
+    std::unordered_map<uint32_t, Instruction*>* preCallSB,
+    std::unique_ptr<BasicBlock>* block_ptr) {
+  (*inst)->ForEachInId(
+      [&postCallSB, &preCallSB, &block_ptr, this](uint32_t* iid) {
+        const auto mapItr = (*postCallSB).find(*iid);
+        if (mapItr == (*postCallSB).end()) {
+          const auto mapItr2 = (*preCallSB).find(*iid);
+          if (mapItr2 != (*preCallSB).end()) {
+            // Clone pre-call same-block ops, map result id.
+            const Instruction* inInst = mapItr2->second;
+            std::unique_ptr<Instruction> sb_inst(inInst->Clone(context()));
+            CloneSameBlockOps(&sb_inst, postCallSB, preCallSB, block_ptr);
+            const uint32_t rid = sb_inst->result_id();
+            const uint32_t nid = this->TakeNextId();
+            get_decoration_mgr()->CloneDecorations(rid, nid);
+            sb_inst->SetResultId(nid);
+            (*postCallSB)[rid] = nid;
+            *iid = nid;
+            (*block_ptr)->AddInstruction(std::move(sb_inst));
+          }
+        } else {
+          // Reset same-block op operand.
+          *iid = mapItr->second;
+        }
+      });
+}
+
+void InstrumentPass::UpdateSucceedingPhis(
+    std::vector<std::unique_ptr<BasicBlock>>& new_blocks) {
+  const auto firstBlk = new_blocks.begin();
+  const auto lastBlk = new_blocks.end() - 1;
+  const uint32_t firstId = (*firstBlk)->id();
+  const uint32_t lastId = (*lastBlk)->id();
+  const BasicBlock& const_last_block = *lastBlk->get();
+  const_last_block.ForEachSuccessorLabel(
+      [&firstId, &lastId, this](const uint32_t succ) {
+        BasicBlock* sbp = this->id2block_[succ];
+        sbp->ForEachPhiInst([&firstId, &lastId](Instruction* phi) {
+          phi->ForEachInId([&firstId, &lastId](uint32_t* id) {
+            if (*id == firstId) *id = lastId;
+          });
+        });
+      });
+}
+
+// Return id for output buffer uint ptr type
+uint32_t InstrumentPass::GetOutputBufferUintPtrId() {
+  if (output_buffer_uint_ptr_id_ == 0) {
+    output_buffer_uint_ptr_id_ = context()->get_type_mgr()->FindPointerToType(
+        GetUintId(), SpvStorageClassStorageBuffer);
+  }
+  return output_buffer_uint_ptr_id_;
+}
+
+uint32_t InstrumentPass::GetOutputBufferBinding() {
+  switch (validation_id_) {
+    case kInstValidationIdBindless: return kDebugOutputBindingBindless;
+    default: assert(false && "unexpected validation id");
+  }
+  return 0;
+}
+
+// Return id for output buffer
+uint32_t InstrumentPass::GetOutputBufferId() {
+  if (output_buffer_id_ == 0) {
+    analysis::TypeManager* type_mgr = context()->get_type_mgr();
```

Comment:
Good question. This needs more clarification comments. The current code assumes just one shared output buffer for all validation passes using synchronized access. Our thinking here was that the buffer would be a stream of output records, each containing an id identifying which validation generated them. But this could be changed with trivial effort. There is very little in this design that locks down how many buffers and which buffers are used by a specific validation pass. Right now we do have the base InstrumentPass header having knowledge about which buffers/bindings are used by which specific passes, but only because some coordination has to be done. Ultimately I am envisioning that, while there will be only one debug descriptor set for all validations, there would be multiple buffers/bindings within that set being used, either jointly or separately, by multiple validation passes. For example, we are imagining adding one or possibly more debug input buffers to assist with with bindless descriptor validation under the VK_EXT_descriptor_indexing extension.

Reply:
Okay, I don't have a strong opinion on how it is done.  It just needs to be documented so the consumer of the data knows where to get it.

Reply:
OK, its documented.